### PR TITLE
CircleCI

### PR DESCRIPTION
**Pull request by leolion3**
Continuous integration using CircleCI.

### CODE REVIEW

**Comment by averysumner** [comment]
Honestly would prefer to add GitHub actions. We used actions internally for a while.

**Comment by leolion3** [comment]
But Github actions would require one to set up shared runners^ CircleCI integrates with Github for free without needing hardware ressources and the pipeline can be viewed in Github

**Comment by MickVermeulen** [comment]
Would it be possible to also do a Windows build with this? Maybe by using https://hub.docker.com/_/microsoft-windows-base-os-images? We've had some issues over the past week where stuff wouldn't build on Linux or Windows and vice versa. 

Also, I'm still not sure I'm a fan of using CircleCI, it forces us to use another third party, I personally have no experience with it as I normally use GitLab CI. Ideally I'd stick with GH to keep everything in one place. Afaik @aronwk-aaron has some more experience with this, maybe he can elaborate.

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+version: 2.1
+
+jobs:
+  build:
+    docker:
+      - image: ubuntu:latest
+    steps:
+      - checkout
+      - run:
+          name: "Set geographic location"
```
Comment (Xiphoseer) [inline .circleci/config.yml:11]: Why is this necessary?

Comment (leolion3) [reply .circleci/config.yml:11]: Checkout checks out the branch and downloads it to the docker instance. Since the docker instance is a fresh environment it has no geographic data set, so when trying to install cmake it will ask you to set a location which you cannot to since you cant interact with the terminal in the pipeline.

**Comment by leolion3** [comment]
> Would it be possible to also do a Windows build with this? Maybe by using https://hub.docker.com/_/microsoft-windows-base-os-images? We've had some issues over the past week where stuff wouldn't build on Linux or Windows and vice versa.
> 
> Also, I'm still not sure I'm a fan of using CircleCI, it forces us to use another third party, I personally have no experience with it as I normally use GitLab CI. Ideally I'd stick with GH to keep everything in one place. Afaik @aronwk-aaron has some more experience with this, maybe he can elaborate.

I myself also only worked with GitlabCI up till last week. I wanted to set up a Github CI but it requires you to set up shared runners, and I don't have the hardware to run them. A quick google search revealed CircleCI as a viable alternative (I had received and ignored dozens of ads about it on Youtube before and thought I'd give it a try). I've used it over the past 4 days and must say it's quite nice.

File: .circleci/config.yml (hunk 1 above)
Comment (Xiphoseer) [reply .circleci/config.yml:11]: > when trying to install cmake it will ask you to set a location
good to know, thanks!

**Comment by Xiphoseer** [comment]
> @leolion3 I myself also only worked with GitlabCI up till last week. I wanted to set up a Github CI but it requires you to set up shared runners, and I don't have the hardware to run them. A quick google search revealed CircleCI as a viable alternative (I had received and ignored dozens of ads about it on Youtube before and thought I'd give it a try). I've used it over the past 4 days and must say it's quite nice.

According to <https://github.com/pricing>, there is some amount of free action time on public repos:

> 2,000 automation minutes/month
> Free for public repositories

**Comment by leolion3** [comment]
> > @leolion3 I myself also only worked with GitlabCI up till last week. I wanted to set up a Github CI but it requires you to set up shared runners, and I don't have the hardware to run them. A quick google search revealed CircleCI as a viable alternative (I had received and ignored dozens of ads about it on Youtube before and thought I'd give it a try). I've used it over the past 4 days and must say it's quite nice.
> 
> According to https://github.com/pricing, there is some amount of free action time on public repos such as this one:
> 
> > 2,000 automation minutes/month
> > Free for public repositories

Oh I didn't know that. Well I mean the CircleCI pipeline is already set up and ready to roll, comes with 6000 minutes/month and you can just login using Github. I mean you could set up Github runners I guess but that would just be additional work.

**Comment by Xiphoseer** [comment]
> Oh I didn't know that. Well I mean the CircleCI pipeline is already set up and ready to roll, comes with 6000 minutes/month and you can just login using Github. I mean you could set up Github runners I guess but that would just be additional work.

You don't actually need to set up those runners (as you would with GitLab for example), GitHub does that automatically. GitHub actions also comes with a lot of tools pre-installed (GCC, Git, CMake, see <https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md>)

**Comment by Xiphoseer** [comment]
> comes with a lot of tools pre-installed

I tried this out yesterday (c.f #248) and it turns out that the GitHub actions linux image has all dependencies we need already pre-installed. How fast is the circle CI runner for a full build @leolion3 ? For comparison, the github actions build takes 8-10 minutes depending on how we invoke it.

**Comment by leolion3** [comment]
@Xiphoseer the last build took 5 minutes and 44 seconds, cmake ran using 8 threads (you get a "big" docker container with 8 cores for DLU
Screenshot 2021-12-17 154442
)

**Review by Xiphoseer — CHANGES_REQUESTED** [review]
Please rebase / recreate / cherry pick this PR onto the main again.

**Comment by averysumner** [comment]
Now that we have GitHub actions, I see no reason to support CircleCI with actions setup. Closing.
